File: 2025/senior3.cpp
#include <bits/stdc++.h>
#define int long long
using namespace std;

signed main() {
	ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int N,M ,Q;
    cin >> N >> M >> Q;
    vector<array<int, 2>> arr;
    for (int i = 0; i < N; i++){
        int a,b;
        cin >> a >> b;
        arr.push_back({a, b});
    }

    unordered_map<int, multiset<int>> vals;
    for (auto i: arr){
        vals[i[0]].insert(i[1]);
    }
    for (auto [k,v]: vals){
        vals[k].insert(-1);
    }

    multiset<int> minimum;  // maintain
    multiset<int> second;
    int total = 0;
    for (auto [k,v]: vals){
        minimum.insert(*--v.end());
        second.insert(*----v.end());
        total += *--v.end();
    }

    for (int q = 0; q <= Q; q++){
        if (q != 0){
            int t,i,x;
            cin >> t >> i >> x;
            i --;

            auto [old_c, old_v] = arr[i];
            if (t == 1) arr[i][0] = x;
            else arr[i][1] = x;
            auto [new_c, new_v] = arr[i];

            minimum.erase(minimum.find(*--vals[old_c].end()));
            second.erase(second.find(*----vals[old_c].end()));

            total -= *--vals[old_c].end();
            vals[old_c].erase(vals[old_c].find( old_v));

            if (old_c != new_c){
                minimum.erase(minimum.find(*--vals[new_c].end()));
                second.erase(second.find(*----vals[new_c].end()));
            }
            total -= *--vals[new_c].end();


            if (old_c != new_c){
                minimum.insert(*--vals[old_c].end());
                second.insert(*----vals[old_c].end());
            }

            total += *--vals[old_c].end();

            vals[new_c].insert(new_v);
            minimum.insert(*--vals[new_c].end());
            second.insert(*----vals[new_c].end());
            total += *--vals[new_c].end();
        }
        cout << total + max(0LL, *--second.end() - *minimum.begin()) << "\n";

    }

}

File: 2025/senior_3.py
from collections import defaultdict
import sys
from bisect import insort, bisect_left

input = sys.stdin.readline
N, M, Q = map(int, input().split())

arr = [list(map(int, input().split())) for _ in range(N)]  # color, pretty

vals = defaultdict(list)  # by color
for a, b in arr:
    vals[a].append(b)

for v in vals:  # padding
    vals[v].append(-1)
    vals[v].sort()

# MAINTAIN THESE
minimum = sorted(([v[-1] for v in vals.values()]))
second = sorted(([v[-2] for v in vals.values()]))
total = sum(v[-1] for v in vals.values())

for _ in range(Q + 1):
    if _ != 0:
        t, i, x = map(int, input().split())
        i -= 1  # 0-index

        # remove old
        old_c, old_v = arr[i]
        if t == 1:  # color of i becomes x
            arr[i][0] = x
        else:  # prettiness of i becomes x
            arr[i][1] = x
        new_c, new_v = arr[i]

        minimum.pop(bisect_left(minimum, vals[old_c][-1]))
        second.pop(bisect_left(second, vals[old_c][-2]))
        total -= vals[old_c][-1]
        vals[old_c].pop(bisect_left(vals[old_c], old_v))

        if old_c != new_c:
            minimum.pop(bisect_left(minimum, vals[new_c][-1]))
            second.pop(bisect_left(second, vals[new_c][-2]))
        total -= vals[new_c][-1]

        # add new
        if old_c != new_c:
            insort(minimum, vals[old_c][-1])
            insort(second, vals[old_c][-2])
        total += vals[old_c][-1]

        insort(vals[new_c], new_v)
        insort(minimum, vals[new_c][-1])
        insort(second, vals[new_c][-2])
        total += vals[new_c][-1]

    print(total + max(0, (second[-1] - minimum[0])))

"""
6 3 0
1 6
2 9
3 4
2 7
3 9
1 3

3 2 2
1 20
2 30
1 10
1 3 2
2 3 25
"""
